File: Java/971.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<Integer> res = new ArrayList<>();
    int i = 0;
    
    public List<Integer> flipMatchVoyage(TreeNode root, int[] voyage) {
        
        if (dfs(root, voyage)) return res;
        
        res.clear();
        res.add(-1);
        return res;
    }
    
    private boolean dfs(TreeNode node, int[] voyage) {
        if (node == null) return true;
        
        if (node.val != voyage[i++]) return false;
        
        if (node.left != null && node.left.val != voyage[i]) {
            res.add(node.val);
            return dfs(node.right, voyage) && dfs(node.left, voyage);
        }
        
        return dfs(node.left, voyage) && dfs(node.right, voyage);
        
    }
}
__________________________________________________________________________________________________
sample 37992 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> flipMatchVoyage(TreeNode root, int[] voyage) {
        List<Integer> result=new ArrayList<>();
        Stack<TreeNode> stack=new Stack<>();
        stack.push(root);
        int i=0;
        while(stack.size()>0)
        {
            TreeNode node=stack.pop();
            if(node==null)
                continue;
            if(node.val!=voyage[i++])
                return Arrays.asList(-1);
            if(node.right!=null && node.right.val==voyage[i])
            {
                if(node.left!=null)
                {
                    result.add(node.val);
                }
                stack.push(node.left);
                stack.push(node.right);
            }
            else
            {
                stack.push(node.right);
                stack.push(node.left);
            }
                
        }
        return result;
    }
}
__________________________________________________________________________________________________

File: Python3/971.py
__________________________________________________________________________________________________
sample 32 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flipMatchVoyage(self, root: TreeNode, voyage: List[int]) -> List[int]:
        voyage.reverse()
        ans = []
        def f(n):
            nonlocal ans, voyage
            sw = 0
            if n is None:
                return True
            elif n.val != voyage[-1]:
                return False
            else:
                voyage.pop()
                if f(n.left) is False and (not ans or ans[-1] > 0):
                    sw += 1
                    ans.append(n.val)
                    n.left, n.right = n.right, n.left
                    
                if sw == 1 and f(n.left) is False:
                    ans.append(-1)
                    return False
                if f(n.right) is False:
                    ans.append(-1)
                    return False
                return True
        ok = f(root)
        return [-1] if not ok or (ans and ans[-1] < 0) or voyage else ans
__________________________________________________________________________________________________
sample 13220 kb submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flipMatchVoyage(self, root: TreeNode, voyage: List[int]) -> List[int]:
        self.flipped = []
        self.i = 0

        def dfs(node):
            if node:
                if node.val != voyage[self.i]:
                    self.flipped = [-1]
                    return
                self.i += 1

                if (self.i < len(voyage) and
                        node.left and node.left.val != voyage[self.i]):
                    self.flipped.append(node.val)
                    dfs(node.right)
                    dfs(node.left)
                else:
                    dfs(node.left)
                    dfs(node.right)

        dfs(root)
        if self.flipped and self.flipped[0] == -1:
            self.flipped = [-1]
        return self.flipped
__________________________________________________________________________________________________
